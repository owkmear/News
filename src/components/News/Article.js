import React from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'
import parse from 'html-react-parser'

class Article extends React.Component {
  static propTypes = {
    imageurl: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    body: PropTypes.string.isRequired
  }

  render() {
    const {
      imageurl,
      title,
      body,
      source_info: { name },
      published_on: publishedOn,
      categories,
      url
    } = this.props
    const timeFromNow = moment.unix(publishedOn).fromNow()
    const categoriesList = categories.split('|').reduce((prev, curr) => [
      prev,
      <span key={prev} className="news__category">
        {' '}
        |{' '}
      </span>,
      curr
    ])
    const imageTitle = `See the full article from ${name}`
    let parsedBody = parse(body)
    if (parsedBody.length > 200) parsedBody = parsedBody.slice(0, 200) + ' ...'

    return (
      <div className="news__article">
        <div className="news__article-image">
          <a href={url} target="_blank" rel="noopener noreferrer nofollow" title={imageTitle}>
            <img src={imageurl} alt={title} />
          </a>
        </div>
        <div>
          <span className="news__name">{name}</span>
          <span className="news__time">{timeFromNow}</span>
          <h3 className="news__title">
            <a href={url} target="_blank" rel="noopener noreferrer nofollow">
              {title}
            </a>
          </h3>
          <p className="news__body">{parsedBody}</p>
          <span className="news__categories-title">Categories: </span>
          <span className="news__categories-list">{categoriesList}</span>
        </div>
      </div>
    )
  }
}

export default Article
