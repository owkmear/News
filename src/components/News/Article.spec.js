import React from 'react'
import { shallow } from 'enzyme'
import Article from './Article'

describe('Article component', () => {
  const props = {
    imageurl: 'https://images.cryptocompare.com/news/cointelegraph/dqkg30ggA0x.png',
    title: '$3B Ponzi Scheme Is Now Allegedly Dumping Bitcoin by the Hundreds',
    body: 'Wallets belonging to PlusToken reportedly contain thousands of Bitcoin and other cryptocurrencies',
    source_info: {
      name: 'CoinTelegraph'
    },
    published_on: 1565863140,
    categories: 'BTC|ETH|Market|Asia',
    url: 'https://cointelegraph.com/news/3b-ponzi-scheme-is-now-allegedly-dumping-bitcoin-by-the-hundreds'
  }

  describe('Article component should render correctly', () => {
    const nextProps = {
      ...props,
      imageurl: 'https://images.cryptocompare.com/news/financemagnates/coMecOic4g0.jpeg',
      title: 'Mastercard Seems to Be All Set to Develop Its Own Crypto Wallet Platform',
      source_info: {
        name: 'CryptoNewsZ'
      },
      body: 'PlusToken, a scam that spread to China and Korea in the past months, accrued vast Bitcoin (BTC) reserves, currently liquidating them in daily tranches.',
      url: 'https://images.cryptocompare.com/news/default/coinspeaker.png'
    }

    const wrapper = shallow(<Article {...nextProps} />)

    it('Render properly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('Should exist', () => {
      expect(wrapper.exists()).toEqual(true)
    })

    it('Should have div', () => {
      expect(wrapper.is('div')).toEqual(true)
    })

    it('Should have class news__article', () => {
      expect(wrapper.hasClass('news__article')).toEqual(true)
    })

    it('Should have news source name', () => {
      expect(wrapper.find('.news__name').text()).toEqual(nextProps.source_info.name)
    })

    it('Should have news text', () => {
      expect(wrapper.find('.news__body').text()).toEqual(nextProps.body)
    })

    it('Should have image', () => {
      expect(wrapper.find('.news__article-image img').at(0).props().src).toEqual(nextProps.imageurl)
      expect(wrapper.find('.news__article-image a').at(0).props().href).toEqual(nextProps.url)
      expect(wrapper.find('img').parent().is('a')).toEqual(true)
    })
  })
})
