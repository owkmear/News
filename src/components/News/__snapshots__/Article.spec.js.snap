// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article component Article component should render correctly Render properly 1`] = `
<div
  className="news__article"
>
  <div
    className="news__article-image"
  >
    <a
      href="https://images.cryptocompare.com/news/default/coinspeaker.png"
      rel="noopener noreferrer nofollow"
      target="_blank"
      title="See the full article from CryptoNewsZ"
    >
      <img
        alt="Mastercard Seems to Be All Set to Develop Its Own Crypto Wallet Platform"
        src="https://images.cryptocompare.com/news/financemagnates/coMecOic4g0.jpeg"
      />
    </a>
  </div>
  <div>
    <span
      className="news__name"
    >
      CryptoNewsZ
    </span>
    <span
      className="news__time"
    >
      2 hours ago
    </span>
    <h3
      className="news__title"
    >
      <a
        href="https://images.cryptocompare.com/news/default/coinspeaker.png"
        rel="noopener noreferrer nofollow"
        target="_blank"
      >
        Mastercard Seems to Be All Set to Develop Its Own Crypto Wallet Platform
      </a>
    </h3>
    <p
      className="news__body"
    >
      PlusToken, a scam that spread to China and Korea in the past months, accrued vast Bitcoin (BTC) reserves, currently liquidating them in daily tranches.
    </p>
    <span
      className="news__categories-title"
    >
      Categories: 
    </span>
    <span
      className="news__categories-list"
    >
      BTC
      <span
        className="news__category"
        key="BTC"
      >
         | 
      </span>
      ETH
      <span
        className="news__category"
        key="BTC,[object Object],ETH"
      >
         | 
      </span>
      Market
      <span
        className="news__category"
        key="BTC,[object Object],ETH,[object Object],Market"
      >
         | 
      </span>
      Asia
    </span>
  </div>
</div>
`;
